src/box/memtx_space.c

```diff
@@ -1473,15 +1473,11 @@ memtx_space_finish_alter(struct space *old_space, struct space *new_space)
 static void
 memtx_space_invalidate(struct space *space)
 {
-	/*
-	 * Abort all concurrent transactions and invalidate space in MVCC
-	 * only if it is enabled. Otherwise, all concurrent transactions
-	 * will be aborted when DDL will be prepared.
-	 */
-	if (memtx_tx_manager_use_mvcc_engine) {
-		memtx_tx_abort_all_for_ddl(in_txn());
+	/* Abort all readers of old scehma. */
+	memtx_tx_abort_all_for_ddl(in_txn());
```

Comment:
This won't work for Vinyl spaces. The reproducer is below.

<details><summary>Vinyl crash</summary>

```lua
local fiber = require('fiber')

box.cfg{}

local s = box.schema.space.create('foo', {engine = 'vinyl'})
s:create_index('pk')
for i = 1, 100 do
    s:insert({i, i * 10})
end

box.error.injection.set('ERRINJ_WAL_DELAY', true)

local f1 = fiber.new(s.rename, s, 'bar')
f1:set_joinable(true)
fiber.sleep(0.1)

box.error.injection.set('ERRINJ_BUILD_INDEX_DELAY', true)

local f2 = fiber.new(s.create_index, s, 'sk', {parts = {2, 'unsigned'}})
f2:set_joinable(true)
fiber.sleep(0.1)

box.error.injection.set('ERRINJ_WAL_DELAY', false)
box.error.injection.set('ERRINJ_WAL_WRITE', true)

local _, err = f1:join()
print('space rename:', err)

box.error.injection.set('ERRINJ_WAL_WRITE', false)
box.error.injection.set('ERRINJ_BUILD_INDEX_DELAY', false)

local _, err = f2:join()
print('create index:', err)

os.exit(0)
```

</details>

Another thing I dislike about this fix is that the `build_index` and `check_format` space methods may not safely use the space object after a yield.

I think we shouldn't release the `AlterSpaceLock` until the transaction is committed or rolled back. To achieve that we'd have to make it reentrant (i.e. allow to lock it multiple times by the same transaction).

Reply:
Thanks! I completely forgot about vinyl 😄
I tried to come up with a fix for it but it always ends up in a situation when we have to call `txn_check_can_continue` after every possible yield in `vinyl_space_build_index`, and vinyl can yield almost everywhere... Moreover, we should test every such place and so on, so it's technically possible not to lock alter while there is another DDL is being committed, but it just looks a bit ugly.

So I introduced `AlterYieldGuard` - it allows current DDL yield, but throws an error when there is another modification of this space being committed. So, this error can be thrown only when building an index or checking space format.